engine/execution/ingestion/stop_control.go

```diff
@@ -3,318 +3,567 @@ package ingestion
+func (s *StopControl) OnBlockExecuted(h *flow.Header) {
 	s.Lock()
 	defer s.Unlock()
 
-	if s.state == StopControlPaused || s.state == StopControlOff {
+	if s.stopBoundary == nil || s.stopped {
 		return
 	}
 
-	if s.stopAfterExecuting == h.ID() {
-		// double check. Even if requested stopHeight has been changed multiple times,
-		// as long as it matches this block we are safe to terminate
-		if h.Height == s.stopHeight-1 {
-			s.stopExecution()
-		} else {
-			s.log.Warn().
-				Msgf(
-					"Inconsistent stopping state. "+
-						"Scheduled to stop after executing block ID %s and height %d, "+
-						"but this block has a height %d. ",
-					h.ID().String(),
-					s.stopHeight-1,
-					h.Height,
-				)
-		}
+	if s.stopBoundary.stopAfterExecuting != h.ID() {
+		return
+	}
+
+	// double check. Even if requested stopHeight has been changed multiple times,
+	// as long as it matches this block we are safe to terminate
+	if h.Height != s.stopBoundary.StopHeight-1 {
+		s.log.Warn().
+			Msgf(
+				"Inconsistent stopping state. "+
+					"Scheduled to stop after executing block ID %s and height %d, "+
+					"but this block has a height %d. ",
+				h.ID().String(),
+				s.stopBoundary.StopHeight-1,
+				h.Height,
+			)
+		return
 	}
+
+	s.stopExecution()
 }
 
 func (s *StopControl) stopExecution() {
-	if s.crash {
-		s.log.Fatal().Msgf(
-			"Crashing as finalization reached requested "+
-				"stop height %d and the highest executed block is (%d - 1)",
-			s.stopHeight,
-			s.stopHeight,
-		)
+	log := s.log.With().
+		Stringer("requested_stop", s.stopBoundary).
+		Uint64("last_executed_height", s.stopBoundary.StopHeight).
+		Stringer("last_executed_id", s.stopBoundary.stopAfterExecuting).
+		Logger()
+
+	s.stopped = true
+	log.Warn().Msg("Stopping as finalization reached requested stop")
+
+	if s.stopBoundary != nil && s.stopBoundary.ShouldCrash {
+		// TODO: crash more gracefully or at least in a more explicit way
+		log.Fatal().Msg("Crashing as finalization reached requested stop")
 		return
 	}
+}
 
-	s.log.Debug().
-		Int8("previous_state", int8(s.state)).
-		Int8("new_state", int8(StopControlPaused)).
-		Msg("StopControl state transition")
+func (s *StopControl) handleVersionBeacon(
+	height uint64,
+) error {
+	if s.nodeVersion == nil {
+		return nil
+	}
 
-	s.state = StopControlPaused
+	if s.versionBeacon != nil && s.versionBeacon.SealHeight >= height {
+		// we already processed this or a higher version beacon
+		return nil
+	}
 
-	s.log.Warn().Msgf(
-		"Pausing execution as finalization reached "+
-			"the requested stop height %d",
-		s.stopHeight,
-	)
+	vb, err := s.versionBeacons.Highest(height)
+	if err != nil {
+		return fmt.Errorf("failed to get highest "+
+			"version beacon for stop control: %w", err)
+	}
 
-}
+	if vb == nil {
+		// no version beacon found
+		// this is unexpected as there should always be at least the
+		// starting version beacon, but not fatal.
+		// It can happen if the node starts before bootstrap is finished.
```

Comment:
By bootstrap, do you mean protocol state bootstrap? That should happen strictly before engines and components are started. It would be great if we could require that bootstrapping is completed strictly before this component is started - then we can fail fast for this edge case rather than quietly continuing in a potentially bad state. 

Also, given that having no version beacon is not expected, I would suggest to change `Highest` to return `storage.ErrNotFound` in this case. Returning this error when an entity is not found in storage is more consistent with other storage interfaces, and makes it clear that we are in an unexpected state.